File: cpp/except2.cpp
#include <iostream>
#include <cstdlib>
#include <exception>

using namespace std;
struct NegException : public exception {
   const char * what () const throw () {
      return "Negative numbers don't have a factorial!";
   }
};

struct BigException : public exception {
   const char * what () const throw () {
      return "The number is too big!";
   }
};

int fact(int i){
  if (i < 0)
    throw NegException();
    if (i > 5)
      throw BigException();

	if (i ==1)
		return 1;
	return i * fact(i-1);
}

int main(int argc, char *argv[])
{

  if (argc < 2) {
    cout << "Please enter an integer value\n";
    exit(0);
  }

  int n = atoi(argv[1]);

  try {
    int f = fact(n);
    cout << "Fact(" << n << ")= " << f << "\n" ;
  }
  catch (exception& e) {
    cout << e.what() << "\n";
  }
}

File: ruby/except2.rb
module Errors
	class NegException < StandardError
		def initialize()
			super("Negative numbers don't have a factorial!")
		end
	end
	class BigException < StandardError
		def initialize()
			super("The number is too big!")
		end
	end
end

def fact(i)
	case
	when i < 0
		raise Errors::NegException.new()
	when i > 5
		raise Errors::BigException.new()
	when i == 0
		1
	else
		i * fact(i-1)
	end
end

if ARGV.length < 1
	print "Please enter an integer value\n"
	exit 0
end

n = ARGV[0].to_i

begin
	f = fact(n)
	puts "Fact(" + n.to_s + ")= " + f.to_s
rescue Errors::NegException => e
	puts e.message + "\n"
rescue Errors::BigException => e
	puts e.message + "\n"
else
	puts "No exception\n"
ensure
	puts "That's all!\n"
end
